### ISSUE TITLE
Implement Insufficient Permissions error / Request Access CTA for KMW tiles

### ISSUE THREAD

**Issue by bethanylang** [google/site-kit-wp#7465]
[## Bug Description

The Error state of the KMW tiles does not correctly implement the "Insufficient Permissions" error where another "Request Access" CTA is displayed instead of the usual "Retry" button CTA.

As reported by @mohitwp in Bug Bash:

> Admin 2 has insufficient permissions for Analytics but KMW tiles showing 'Request access' button only for Search Console and other KMW tiles related to Analytics not appearing with 'Request Access' button. 

![image](https://github.com/google/site-kit-wp/assets/12189958/e56baef7-76c2-44e1-9adc-b1622592107b)

![image](https://github.com/google/site-kit-wp/assets/12189958/0a4a8163-820d-43d3-82e5-b45665282873)

As reported by @mohitwp during Bug Bash:

In case of insufficient permission. Request access button position is not correct.

![image](https://github.com/google/site-kit-wp/assets/12189958/6d2b5f27-feee-45cd-b498-45729e4fce07)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### Analytics Permissions

* When a logged-in admin does not have access to the connected **_GA4 property_**, any Key Metric widget tiles that rely on GA4 data should render the "Data loading failed" error tiles designed in #7310 as per this [Figma mock](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/Key-Metrics?type=design&node-id=5409-80974&mode=design&t=9lwlYoTsu7f0jhhu-0), but with the following changes:
  * The error message within should be changed to `"Insufficient permissions"` (instead of `"Data loading failed"`)
  * The `"Request Access"` CTA button should be added instead of the `"Retry"` button. Clicking this button should take you to the GA4 Missing Permissions popup in the GA4 Admin console as implemented for other widgets in #7492. ([See this screenshot for the page the user should be taken to.](https://github.com/google/site-kit-wp/assets/1721319/37bc19c8-c362-4e44-a2c3-bef9746d304a))
  * The `"Retry didn't work?"` text should be changed to `"Trouble getting access?"`
  * The `"Get help"` link below the error message should link to `https://sitekit.withgoogle.com/documentation/troubleshooting/dashboard/#insufficientpermissionsanalytics`.

### Search Console Permissions

* When a logged-in admin does not have access to the connected **_Search Console property_**, any Key Metric widget tiles that rely on Search Console data should render the "Data loading failed" error tiles designed in #7310 as per this [Figma mock](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/Key-Metrics?type=design&node-id=5409-80974&mode=design&t=9lwlYoTsu7f0jhhu-0), but with the following changes:
  * The error message within should be changed to `"Insufficient permissions"` (instead of `"Data loading failed"`)
  * The `"Retry didn't work?"` text should be changed to `"Trouble getting access?"`
  * The `"Get help"` link below the error message already points to the correct page. This should not change.
  * The `"Request Access"` CTA button should take the user to the Search Console "not verified" screen as it currently does for other Search Console widgets. ([See this screenshot for the page the user should be taken to](https://github.com/google/site-kit-wp/assets/1721319/9d2ad037-2748-4104-b9a8-d20fef8885d2).)

## Implementation Brief

NOTE TO IB author: See the [comment below](https://github.com/google/site-kit-wp/issues/7465#issuecomment-1693186409) for more context on solving this.

* Once #7492 is merged, the `getServiceEntityAccessURL` selector will be available to get the `Request Access` URL for the `analytics-4` module.

In `assets/js/components/ReportErrorActions.js`:

* Add a new prop named `getHelpText` of type string.
* If the `getHelpText` prop is provided, render the `Get help` link with the provided text, similar to how `Retry didn’t work?` and `Get help` are rendered.
* Otherwise, render the `Get help` link as it is currently.

In `assets/js/components/KeyMetrics/MetricTileError.js`:

* Use the `isInsufficientPermissionsError` utility function to check for insufficient permissions errors.
* Depending on the error type, dynamically set the `title` prop of the `CTA` component to either `Insufficient permissions` or `Data loading failed`.
* If the error is an insufficient permissions error, pass `"Trouble getting access?"` as the `getHelpText` prop to the `ReportErrorActions` component.

In `assets/sass/components/key-metrics/_googlesitekit-km-widget-tile.scss`:

* Move the `.googlesitekit-cta-link--external` class styles one level up to the `.googlesitekit-report-error-actions` class. Currently, it is nested to `.googlesitekit-error-retry-text`. This will ensure the `Get help` link is styled correctly for both the `Retry` and `Request Access` buttons.
* Extend the styles from the `.googlesitekit-report-error-actions.button` to also apply to `a` tag, ensuring consistent styling for both the `Retry` and `Request Access` buttons. This is because the `Retry` button is rendered as a `button`; however, `Request Access` is rendered as an `a` tag.

#### Stories

* Add new stories in all Key Metrics widget story files to depict the "Insufficient permissions" error state.

**Note:** A new error code, `analytics-4_insufficient_permissions`, will redirect to [this documentation page](https://sitekit.withgoogle.com/documentation/troubleshooting/dashboard/#insufficientpermissionsanalytics), according to the error states mapping sheet. This feature is not yet in place, but no additional code changes should be needed here. Please refer to this [comment](https://github.com/google/site-kit-wp/issues/7465#issuecomment-1705082295).

### Test Coverage

* Fix any broken tests that arise due to the modifications.
* Update any VRT images that may be affected by the changes.

## QA Brief

In order to simulate the Insufficient Permissions error, API requests to applicable KMW endpoints will need to be intercepted and said responses altered in a manner that forces an error. The Chrome `Tweak` extension (as used in prior QA tests) should be leveraged. Otherwise, follow the "Steps To Reproduce" in #7492 to get the actual error state.

**Steps to follow:**

- Ensure that the `userInput` feature flag is enabled.
- Ensure that KMW tiles are configured using the wizard or via the SK settings.
- Open the Tweak extension in Chrome and add the following values to the fields:
    - URL: `^(.*)\/wp-json\/google-site-kit\/v1\/modules\/(.*)$` (Enable Regex matching via the `.*` button to the right of the field)
    - HTTP Method: `GET`
    - Status Code: `403`
    - Delay By: `0s`
    - Response Payload:
```
{
  "code": 403,
  "message": "Insufficient Permissions Test Error",
  "data": {
    "status": 403,
    "reason": "insufficientPermissions"
  }
}
```
- Ensure that the extension and rule are both enabled.
- The final configuration should look as follows:

![Site Kit by Google Dashboard ‹ sitekit 10uplabs com — WordPress 2023-09-07 at 12 02 26 AM](https://github.com/google/site-kit-wp/assets/72872375/40cfd35f-d88f-4763-8571-e6f6ed5cbaa9)

- Via the Chrome web inspector panel, within the `Application` tab, clear the following:
   - Local Storage for the site on which QA is being undertaken.
   - Session Storage for the site on which QA is being undertaken.
   - Cookies for the site on which QA is being undertaken.
- Log into the site and navigate to the SK dashboard, the default widgets should have the correct insufficient permissions error tile and CTA as per below (this will vary based on your chosen KMW's):

![Site Kit by Google Dashboard ‹ sitekit 10uplabs com — WordPress 2023-09-08 at 2 23 11 PM](https://github.com/google/site-kit-wp/assets/72872375/1593f023-1c41-49d3-972c-a13b089ec55e)

- Next, paste the following JS snippet into the console within Chrome web inspector, this will enable all KMW to be displayed:

```
await googlesitekit.data.dispatch('core/user').setKeyMetricsSetting('widgetSlugs', ['kmAnalyticsAdSenseTopEarningContent', 'kmAnalyticsEngagedTrafficSource', 'kmAnalyticsLoyalVisitors', 'kmAnalyticsNewVisitors', 'kmAnalyticsPopularContent', 'kmAnalyticsPopularProducts', 'kmAnalyticsTopCities', 'kmAnalyticsTopConvertingTrafficSource', 'kmAnalyticsTopCountries', 'kmAnalyticsTopTrafficSource', 'kmSearchConsolePopularKeywords'] );
await googlesitekit.data.dispatch('core/user').saveKeyMetricsSettings();
```

- This should render all available KMW tiles and they should all error with the correct Insufficient Permissions error tile and accompanying CTA's, as per the screenshot below:

![Site Kit by Google Dashboard ‹ sitekit 10uplabs com — WordPress 2023-09-08 at 2 27 09 PM](https://github.com/google/site-kit-wp/assets/72872375/e9d07970-c739-4b6c-8db9-1cb375c7cffe)

## Changelog entry

* Show a specific error message and a "Request access" CTA for Key Metrics widget tiles when they encounter a permissions error.

**Comment by marrrmarrr** [comment]
@jimmymadon agree that simply hiding the whole section without any explanation would be confusing for people. 
I'm not sure if they will connect the message "Data loading failed" with permissions issues, in which case having a "request access" CTA could also be confusing? (i.e. if the it's the loading that failed, "retry" makes more sense, but actually retrying won't do anything here, since they're having a different problem)

Is it possible to replace "Data loading failed" with "Insufficient permissions" here? Then the "request access" CTA would be more coherent with the error (and we could go with Option 1 on your list).

**Comment by tofumatt** [comment]
I'm finding this issue difficult to follow, I think the wording of the problem isn't very clear, but additionally I think there's multiple things being reported.

The screenshots in the issue breaking up the context and seemingly referring to different things is making the ACs here unclear.


-----


I think the ACs here are getting at this:

* For Key Metrics Widgets using GA4 data, if the user doesn't have permissions to access the Analytics property, we should show a "Request Access" error tile instead of "Data loading error".
* The button for this error should be moved down to match the Figma design's layout

This is correct, right? I think it'd be better to go through the ACs here again to simplify them and focus the ACs.

**Comment by tofumatt** [comment]
Made a few tweaks and discussed them with @jimmymadon, this is good to go to IB now 🙂

**Comment by techanvil** [comment]
Hey @hussain-t, this IB is looking good. I do have a couple of points:

I realise this was just following the AC, but the title should be "Insufficient permissions", rather than "Insufficient Permissions", for casing consistency with both the alternative title "Data loading failed" as well as the "Insufficient permissions in {module name}" error message seen in the other Dashboard widgets, and also per Mariya's [comment](https://github.com/google/site-kit-wp/issues/7465#issuecomment-1692667753) above. I've tweaked the corresponding string in the AC.

It's also worth adding a note to the IB to clarify the point that there _is_ now going to be an `analytics-4_insufficient_permissions` error code redirect to https://sitekit.withgoogle.com/documentation/troubleshooting/dashboard/#insufficientpermissionsanalytics according to the error states mapping [sheet](https://docs.google.com/spreadsheets/d/12x39d7j0GNqha0m7s4-Y3P5SwK9wwgxxxDnWp3svSKg/edit?resourcekey=0-IngulwDvPmI4SEq2g-f4rA#gid=0), although this is not yet in place - meaning that the AC for the "get help" link can't be fully tested until this is in place, but no additional code changes should be needed here (i.e. we should not need to link to the `analytics_insufficient_permissions` error as discussed in [this comment](https://github.com/google/site-kit-wp/issues/7465#issuecomment-1693186409) above and the related [Asana task](https://app.asana.com/0/1185360131189553/1205277079738801/f)). cc @jamesozzie 

Additionally, in retrospect it's worth adding a similar point to the QAB for https://github.com/google/site-kit-wp/issues/7492 as the switch of the module slug from `analytics` to `analytics-4` in that issue will have introduced a temporary regression in the "Get help" links for the related widgets.

**Comment by hussain-t** [comment]
Thanks, @techanvil. I've updated the IB. It's back to you for another review.

**Comment by techanvil** [comment]
Thanks @hussain-t! IB LGTM :white_check_mark:

**Comment by wpdarren** [comment]
### QA Update:  ✅ 

**Verified:**

- Went through the steps in the QA (thank you @10upsimon for the fantastic instructions as always!) 
  - All available KMW tiles are rendered and they show all error with the correct Insufficient Permissions error tile and accompanying CTA's, as per the screenshot below:
  - Checked that the CTAs opened up in search console or analytics to give permission. 
  - Checked that the 'Get help' links continued to function and the user is sent to the correct help page. 
  - Checked the UX/UI on desktop and smaller screen sizes like tablet and mobile. 

![image](https://github.com/google/site-kit-wp/assets/73545194/a71fbf54-5e1d-49e4-b04c-5d20b7373973)
